tools/sklearn/model_prediction.py

```diff
@@ -137,54 +137,62 @@ def main(inputs, infile_estimator, outfile_predict,
             ref_genome_path=ref_seq, vcf_path=vcf_path, **options)
 
         pred_data_generator.fit()
-
-        preds = estimator.model_.predict_generator(
-            pred_data_generator.flow(batch_size=32),
-            workers=N_JOBS,
-            use_multiprocessing=True)
-
-        if preds.min() < 0. or preds.max() > 1.:
-            warnings.warn('Network returning invalid probability values. '
-                          'The last layer might not normalize predictions '
-                          'into probabilities '
-                          '(like softmax or sigmoid would).')
-
-        if params['method'] == 'predict_proba' and preds.shape[1] == 1:
-            # first column is probability of class 0 and second is of class 1
-            preds = np.hstack([1 - preds, preds])
-
-        elif params['method'] == 'predict':
-            if preds.shape[-1] > 1:
-                # if the last activation is `softmax`, the sum of all
-                # probibilities will 1, the classification is considered as
-                # multi-class problem, otherwise, we take it as multi-label.
-                act = getattr(estimator.model_.layers[-1], 'activation', None)
-                if act and act.__name__ == 'softmax':
-                    classes = preds.argmax(axis=-1)
-                else:
-                    preds = (preds > 0.5).astype('int32')
+        variants = np.array(pred_data_generator.variants)
+        # predict 1600 sample at once then write to file
+        gen_flow = pred_data_generator.flow(batch_size=1600)
+
+        file_writer = open(outfile_predict, 'w')
+        header_row = '\t'.join(['chrom', 'pos', 'name', 'ref',
+                                'alt', 'strand'])
+        file_writer.write(header_row)
+        header_done = False
+
+        steps_done = 0
+
+        # TODO: multiple threading
+        while steps_done < len(gen_flow):
+            index_array = next(gen_flow.index_generator)
+            batch_X = gen_flow._get_batches_of_transformed_samples(
+                index_array)
+
+            if params['method'] == 'predict':
+                batch_preds = estimator.predict(
```

Comment:
Why a `batch_predictor` being used? Because the data for prediction may not fit in memory at once while predicting?

Reply:
All Keras prediction API are built on batches. Data_generator kicks in here to perform data conversion (from vcf info to 2-D array) as prediction goes. Basically, to be MEM efficient, as the tensorflow itself is a MEM monster.